bliss/metrics.py

```diff
@@ -36,128 +39,80 @@ def __init__(
+        self.add_state("detection_tp", default=torch.tensor(0), dist_reduce_fx="sum")
+        self.add_state("detection_fp", default=torch.tensor(0), dist_reduce_fx="sum")
         self.add_state("avg_distance", default=torch.tensor(0.0), dist_reduce_fx="sum")
         self.add_state("total_true_n_sources", default=torch.tensor(0), dist_reduce_fx="sum")
-        self.add_state("total_correct_class", default=torch.tensor(0), dist_reduce_fx="sum")
-        self.add_state("conf_matrix", default=torch.tensor([[0, 0], [0, 0]]), dist_reduce_fx="sum")
+        self.add_state("gal_tp", default=torch.tensor(0), dist_reduce_fx="sum")
+        self.add_state("gal_fp", default=torch.tensor(0), dist_reduce_fx="sum")
+        self.add_state("gal_fn", default=torch.tensor(0), dist_reduce_fx="sum")
+        self.add_state("gal_tn", default=torch.tensor(0), dist_reduce_fx="sum")
 
     # pylint: disable=no-member
     def update(self, true: FullCatalog, est: FullCatalog) -> None:  # type: ignore
-        """Update the internal state of the metric including tp, fp, total_true_n_sources, etc."""
+        """Update the internal state of metrics including tp, fp, total_coadd_gal_matches, etc."""
         assert true.batch_size == est.batch_size
 
         count = 0
-        desc = "Detection Metric per batch"
+        desc = "Bliss Metric per batch"
         for b in tqdm(range(true.batch_size), desc=desc, disable=self.disable_bar):
             ntrue, nest = true.n_sources[b].int().item(), est.n_sources[b].int().item()
             tlocs, elocs = true.plocs[b], est.plocs[b]
-            if ntrue > 0 and nest > 0:
-                mtrue, mest, dkeep, avg_distance = match_by_locs(tlocs, elocs, self.slack)
-                tp = len(elocs[mest][dkeep])  # n_matches
-                true_galaxy_bools = true["galaxy_bools"][b][mtrue][dkeep]
-                tp_gal = true_galaxy_bools.bool().sum()
-                fp = nest - tp
-                assert fp >= 0
-                self.tp += tp
-                self.tp_gal += tp_gal
-                self.fp += fp
-                self.avg_distance += avg_distance
-                self.total_true_n_sources += ntrue  # type: ignore
-                count += 1
+            tgbool, egbool = true["galaxy_bools"][b].reshape(-1), est["galaxy_bools"][b].reshape(-1)
+            mtrue, mest, dkeep, avg_distance = match_by_locs(tlocs[:ntrue], elocs[:nest], self.slack)
+            detection_tp = len(elocs[mest][dkeep])
+            detection_fp = nest - detection_tp
+            tgbool = tgbool[mtrue][dkeep].reshape(-1)
+            egbool = egbool[mest][dkeep].reshape(-1)
+            assert fp >= 0
+            self.detection_tp += detection_tp
+            self.detection_fp += detection_fp
+            self.avg_distance += avg_distance
+            self.total_true_n_sources += ntrue  # type: ignore
+            conf_matrix = confusion_matrix(tgbool.cpu(), egbool.cpu(), labels=[1, 0])
```

Comment:
Confusion_matrix converts the iterable inputs to Numpy arrays, so **_tgbool_** and **_egbool_** have to be allocated on the CPU since Numpy arrays are not supported on the GPU.